lib/nodes/orchestra/ddf.cpp

```diff
@@ -0,0 +1,410 @@
+/* OPAL-RT Orchestra Data Definition File (DDF)
+ *
+ * Author: Steffen Vogel <steffen.vogel@opal-rt.com>
+ *
+ * SPDX-FileCopyrightText: 2025, OPAL-RT Germany GmbH
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+extern "C" {
+#include <RTAPI.h>
+}
+
+#include <villas/nodes/orchestra/ddf.hpp>
+#include <villas/nodes/orchestra/error.hpp>
+#include <villas/nodes/orchestra/signal.hpp>
+#include <villas/sample.hpp>
+#include <villas/signal.hpp>
+
+using namespace villas::node;
+using namespace villas::node::orchestra;
+
+static std::vector<std::string> split(const std::string &path,
+                                      char delimiter = '/') {
+  std::vector<std::string> components;
+  size_t start = 0, end = 0;
+  while ((end = path.find(delimiter, start)) != std::string::npos) {
+    if (end != start) {
+      components.push_back(path.substr(start, end - start));
+    }
+    start = end + 1;
+  }
+  if (start < path.size()) {
+    components.push_back(path.substr(start));
+  }
+  return components;
+}
+
+void DataItem::toXml(xmlNode *parent, bool withDefault) const {
+  xmlNode *item = xmlNewChild(parent, nullptr, BAD_CAST "item", nullptr);
+  xmlNewProp(item, BAD_CAST "name", BAD_CAST name.c_str());
+  xmlNewChild(item, nullptr, BAD_CAST "type",
+              BAD_CAST signalTypeToString(type).c_str());
+  xmlNewChild(item, nullptr, BAD_CAST "length",
+              BAD_CAST fmt::format("{}", length).c_str());
+
+  if (withDefault) {
+    std::string defaultValueStr;
+    switch (type) {
+    case orchestra::SignalType::BOOLEAN:
+      defaultValueStr = defaultValue ? "yes" : "no";
+      break;
+
+    default:
+      defaultValueStr = fmt::format("{}", defaultValue);
+      break;
+    }
+
+    xmlNewChild(item, nullptr, BAD_CAST "default",
+                BAD_CAST defaultValueStr.c_str());
+  }
+}
+
+void BusItem::toXml(xmlNode *parent, bool withDefault) const {
+  xmlNode *node = xmlNewChild(parent, NULL, BAD_CAST "item", NULL);
+  xmlNewProp(node, BAD_CAST "name", BAD_CAST name.c_str());
+  xmlNewChild(node, NULL, BAD_CAST "type", BAD_CAST "bus");
+
+  for (const auto &p : items) {
+    auto &item = p.second;
+    item->toXml(node, withDefault);
+  }
+}
+
+DataItem *BusItem::upsertItem(std::vector<std::string> pathComponents,
```

Comment:
could use a more descriptive name. What is upsert?

Reply:
Update is a commonly used term in data base to describe a transaction which either **up**dates or **in**serts an entry depending on wether it already exists.